Issue 3871 configure bin hashes

s1gse9v opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

I have implemented the functionality requested in #3871. `iT` now computes and shows the algorithms specified in `bin.hashes.default`. 

...

**Test plan**

I have tested using `test/db/cmd/cmd_i`. I'm having trouble running the remaining tests however (even on `dev`) so more testing is probably required.

...

**Closing issues**

closes #3871 

...

s1gse9v committed:
Add bin.hashes.default config option

s1gse9v committed:
Make rz_bin_file_compute_hashes use bin.hashes.default

s1gse9v committed:
Fix null pointer deref in rz_cmd_info_hashes_handler and add spport for hash lists of different lengths

s1gse9v committed:
Formatting

s1gse9v committed:
Add test iT3 in db/cmd/cmd_i for new functionality

wargio commented on librz/bin/bfile.c:
```diff
@@ -426,43 +426,47 @@ RZ_API RZ_OWN RzPVector /*<RzBinFileHash *>*/ *rz_bin_file_compute_hashes(RzBin
-	if (!rz_hash_cfg_init(md)) {
-		goto rz_bin_file_compute_hashes_bad;
-	}
+	if (hash_algos_iter) {
```
i would change this as, if `hash_algos_iter` is `NULL` then goto `rz_bin_file_compute_hashes_bad`.
Maybe also throw an error via `RZ_LOG_ERROR`

wargio commented on librz/core/cconfig.c:
```diff
@@ -2450,6 +2450,68 @@ static bool cb_binverbose(void *user, void *data) {
+	free(core->bin->default_hashes);
+	if (*node->value) {
+		char *delim = ",";
+		char *value = strdup(node->value);
+		if (!value) {
+			rz_sys_perror("strdup");
+			return false;
+		}
+		RzList *algo_list = rz_list_newf((RzListFree)free);
+		char *token = strtok(value, delim);
+		while (token) {
+			for (int i = 0; i < RZ_ARRAY_SIZE(algos); i++) {
+				if (algos[i] != NULL && !strcmp(token, algos[i])) {
+					char *hash = strdup(algos[i]);
+					if (hash) {
+						rz_list_append(algo_list, hash);
+					} else {
+						rz_sys_perror("strdup");
+						rz_list_free(algo_list);
+						free(value);
+						return false;
+					}
+					algos[i] = NULL;
+					break;
+				} else if (i == RZ_ARRAY_SIZE(algos) - 1) {
+					RZ_LOG_ERROR("core: bin.hashes.default: invalid or duplicate value: %s\n", node->value);
+					rz_list_free(algo_list);
+					free(value);
+					return false;
+				}
+			}
+			token = strtok(NULL, delim);
+		}
+		free(value);
+		core->bin->default_hashes = algo_list;
+		return true;
+	} else {
+		core->bin->default_hashes = NULL;
+		return true;
+	}
```
This whole thing can be simplified by doing
```c
core->bin->default_hashes = rz_str_split_duplist(node->value, ",", true);
if (!core->bin->default_hashes) core->bin->default_hashes = rz_list_new();
```
Also there is no need to force those algos.

wargio commented on librz/core/cconfig.c:
```diff
@@ -3245,6 +3307,10 @@ RZ_API int rz_core_config_init(RzCore *core) {
 	SETBPREF("bin.classes", "true", "Load classes from rbin on startup");
 	SETCB("bin.verbose", "false", &cb_binverbose, "Show RzBin warnings when loading binaries");
+	SETCB("bin.hashes.default", "md5,sha1,sha256,crc32,entropy", &cb_binhashesdefault, "Select hash algorithms");
+	n = NODECB("bin.hashes.default", "md5,sha1,sha256,crc32,entropy", &cb_binhashesdefault);
+	SETDESC(n, "Select hashing algorithms");
+	SETOPTIONS(n, "md5", "sha1", "sha256", "crc32", "entropy", "blake3", "sm3", NULL);
```
remove this

wargio reviewed (changes requested):
nice! some changes are needed

s1gse9v committed:
Fix test/integration/test_bin.c

* Because the hashes computed for a given bin are now set by
  cb_binhashes in bin->default_hashes, they have to be set manually in
  test_bin.c

wargio commented on test/integration/test_bin.c:
```diff
@@ -8,6 +8,8 @@
 
 bool test_rz_bin(void) {
 	RzBin *bin = rz_bin_new();
+	const char* default_algos[] = {"md5", "sha1", "sha256", "crc32", "entropy"};
+	bin->default_hashes = rz_list_new_from_array(default_algos, sizeof(default_algos)/sizeof(default_algos[0]));
```
```suggestion
	const char* default_algos[] = {"md5", "sha1", "sha256", "crc32", "entropy"};
	bin->default_hashes = rz_list_new_from_array(default_algos, RZ_ARRAY_SIZE(default_algos));
```

s1gse9v replied:
Thanks for the feedback, I have included your changes in my last commit. Let me know if anything else need changing.

s1gse9v committed:
Add requested changes and fix tests

s1gse9v committed:
Fix formatting

wargio commented on librz/bin/bfile.c:
```diff
@@ -428,45 +428,47 @@ RZ_API RZ_OWN RzPVector /*<RzBinFileHash *>*/ *rz_bin_file_compute_hashes(RzBin
-
-		if (!rz_hash_cfg_init(md)) {
+	if (!hash_algos_iter) {
+		RZ_LOG_ERROR("bin: default hashes: no default hashes configured");
```
```suggestion
		RZ_LOG_ERROR("bin: default hashes: no default hashes configured\n");
```

wargio commented on librz/bin/bfile.c:
```diff
@@ -428,45 +428,47 @@ RZ_API RZ_OWN RzPVector /*<RzBinFileHash *>*/ *rz_bin_file_compute_hashes(RzBin
+		RZ_LOG_ERROR("bin: default hashes: no default hashes configured");
+		goto rz_bin_file_compute_hashes_bad;
+	}
+	while (hash_algos_iter) {
```
This loop is wrong because is an infinite loop... use the normal looping for RzList as shown below.

```c
RzListIter *iter = NULL;
const char *algo = NULL;
rz_list_foreach (bin->default_hashes, iter, algo) {
// call rz_hash_cfg_configure(md, algo) ...
}
```

wargio commented on librz/bin/bfile.c:
```diff
@@ -428,45 +428,47 @@ RZ_API RZ_OWN RzPVector /*<RzBinFileHash *>*/ *rz_bin_file_compute_hashes(RzBin
-			if (!add_file_hash(md, algo, file_hashes)) {
-				goto rz_bin_file_compute_hashes_bad;
-			}
+	while (hash_algos_iter) {
```
same here. wrong loop.

wargio commented on librz/core/cconfig.c:
```diff
@@ -2458,58 +2458,9 @@ static bool cb_binhashesdefault(void *user, void *data) {
-		return true;
-	}
+	core->bin->default_hashes = rz_str_split_duplist(node->value, ",", true);
+	if (!core->bin->default_hashes) core->bin->default_hashes = rz_list_new();
```
```suggestion
	if (!core->bin->default_hashes) {
		core->bin->default_hashes = rz_list_new();
	}
```

wargio commented on test/integration/test_bin.c:
```diff
@@ -9,7 +9,7 @@
 	RzBin *bin = rz_bin_new();
 	const char* default_algos[] = {"md5", "sha1", "sha256", "crc32", "entropy"};
-	bin->default_hashes = rz_list_new_from_array(default_algos, sizeof(default_algos)/sizeof(default_algos[0]));
+	bin->default_hashes = rz_list_new_from_array((void**)default_algos, RZ_ARRAY_SIZE(default_algos));
```
```suggestion
	bin->default_hashes = rz_list_new_from_array((const void**)default_algos, RZ_ARRAY_SIZE(default_algos));
```

wargio reviewed (changes requested):
please add also a test under `db/cmd` where you modify `bin.hashes.default` and print the hashes.

s1gse9v committed:
Extend test/db/cmd/cmd_i

s1gse9v committed:
Implement requested changes

s1gse9v committed:
Formatting

wargio reviewed (commented):
LGTM, just run `./sys/clang-format.py` (requires to install first clang-format 16 https://github.com/rizinorg/rizin/blob/dev/DEVELOPERS.md#code-style)

wargio reviewed (approved):
LGTM. awesome job

XVilka committed:
Merge branch 'dev' into issue-3871-configure-bin-hashes
